src/shims/unix/linux/epoll.rs

```diff
@@ -378,39 +379,105 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
+        // is not close after the thread unblocks.
+        let weak_epfd = epfd.downgrade();
+
         let mut binding = epfd.borrow_mut();
         let epoll_file_description = &mut binding
             .downcast_mut::<Epoll>()
             .ok_or_else(|| err_unsup_format!("non-epoll FD passed to `epoll_wait`"))?;
 
         let binding = epoll_file_description.get_ready_list();
+        let ready_list = binding.borrow_mut();
+
+        if ready_list.is_empty() {
+            if timeout == 0 {
+                // Non-blocking with no notification returned.
+                this.write_scalar(Scalar::from_i32(0), &dest)?;
+                return Ok(());
+            } else {
+                // Blocking
+                let duration = Duration::from_millis(timeout.into());
+                drop(ready_list);
+                this.block_thread(
+                    BlockReason::Epoll,
+                    Some((TimeoutClock::Monotonic, TimeoutAnchor::Relative, duration)),
+                    callback!(
+                        @capture<'tcx> {
+                            epfd_value: i32,
+                            weak_epfd: WeakFileDescriptionRef,
+                            dest: MPlaceTy<'tcx>,
+                            event: MPlaceTy<'tcx>,
+                        }
+                        @unblock = |this| {
+                            this.blocking_epoll_callback(epfd_value, weak_epfd, &dest, &event)?;
+                            Ok(())
+                        }
+                        @timeout = |this| {
+                            // No notification after blocking timeout.
+                            this.write_scalar(Scalar::from_i32(0), &dest)?;
+                            Ok(())
+                        }
+                    ),
+                );
+            }
+        } else {
+            // Immediately return epoll notification without blocking.
+
+            // Attempt to prevent epfd.borrow_mut in blocking_epoll_callback causing ICE.
+            drop(ready_list);
+            this.blocking_epoll_callback(epfd_value, weak_epfd, &dest, &event)?;
+            return Ok(());
+        }
+
+        Ok(())
+    }
+
+    /// Callback function after epoll_wait unblocks
+    fn blocking_epoll_callback(
+        &mut self,
+        epfd_value: i32,
+        weak_epfd: WeakFileDescriptionRef,
+        dest: &MPlaceTy<'tcx>,
+        event: &MPlaceTy<'tcx>,
+    ) -> InterpResult<'tcx> {
+        let this = self.eval_context_mut();
+
+        let Some(epfd) = weak_epfd.upgrade() else {
+            throw_unsup_format!("epoll FD {epfd_value} is closed while blocking.")
+        };
+
+        //TODO: What if, there is another thread that make changes to the global fd table here?
+        // For example close the current epfd, then insert a new one inside?
+
+        let binding = epfd.borrow_mut().downcast_mut::<Epoll>().unwrap().get_ready_list();
```

Comment:
I am currently trying to resolve borrow_mut error that happens to this line. I tried dropping ``epfd``or ``binding`` before calling ``blocking_epoll_callback`` but it couldn't be done because they are all borrowed. So I proceed to try dropping ``ready_list``, the ``borrow_mut`` error still persists.